spec/weather_retriever_spec.rb

```diff
@@ -0,0 +1,224 @@
+        end
+      end
+
+      context 'when database is empty' do
+        it 'calls the client' do
+          set_up_retriever.run
+
+          expect(open_weather_stub).to have_been_made.once
+        end
+
+        it 'augments the database' do
+          set_up_retriever.run
+
+          result = sql.query "SELECT * FROM weather WHERE name = 'London'" do |rows|
+            rows.next_hash.transform_keys(&:to_sym)
+          end
+
+          expect(result).to eq(todays_weather)
+        end
+
+        it 'returns updated weather' do
+          set_up_retriever.run
+
+          expect(presenter).to have_received(:collect_weather).with(todays_weather)
+        end
+
+        it 'has no client or database errors' do
+          set_up_retriever.run
+
+          expect(presenter).not_to have_received(:log_error)
+        end
+      end
+
+      context 'when database has outdated data' do
+        let(:two_days_ago) { Time.utc(today.year, today.month, today.day - 2).to_i }
+
+        before do
+          create_table
+          setup_name_index
+          add_weather_row(two_days_ago)
+        end
+
+        it 'calls the client' do
+          set_up_retriever.run
+
+          expect(open_weather_stub).to have_been_made.once
+        end
+
+        it 'augments the database' do
+          set_up_retriever.run
+
+          result = sql.query "SELECT * FROM weather WHERE name = 'London'" do |rows|
+            rows.next_hash.transform_keys(&:to_sym)
+          end
+
+          expect(result).to eq(todays_weather(id: 2))
+        end
+
+        it 'returns updated weather' do
+          set_up_retriever.run
+
+          expect(presenter).to have_received(:collect_weather).with(todays_weather(id: 2))
+        end
+
+        it 'has no client or database errors' do
+          set_up_retriever.run
+
+          expect(presenter).not_to have_received(:log_error)
+        end
+      end
+    end
+
+    context 'when failure' do
+      before { open_weather_stub(status: 404) }
+
+      it 'returns no weather' do
+        set_up_retriever.run
+
+        expect(presenter).to have_received(:collect_weather).with({})
+      end
+
+      it 'does not augment database' do
+        set_up_retriever.run
+
+        result = sql.query("SELECT * FROM weather WHERE name = 'London'", &:none?)
+
+        expect(result).to be true
+      end
+
+      it 'logs the error if client fails' do
+        set_up_retriever.run
+
+        expect(presenter).to have_received(:log_error).once.with(
+          "I seemed to have lost the weather API!\n It might be because an invalid city was entered."
+        )
+      end
+
+      it 'logs the error when response data is bad' do
+        bad_data = { foo: 'bar' }
+        open_weather_stub(status: 200, body: bad_data.to_json)
+
+        set_up_retriever.run
+
+        expect(presenter).to have_received(:log_error).once.with('NOT NULL constraint failed: weather.name')
+      end
```

Comment:
Is this error in the test coming from the local setup with status 200 and bad data or is it coming from the global setup with status 404 that you are running before every test?

Observe that since you are running the setup with 404 before the test runs, when you run `set_up_retriever.run` it will run it with the first setup (404), and only if you ran `set_up_retriever.run` a second time will it use the local setup with status 200.

So in this case it's better to run `open_weather_stub(status: 404)` locally in every test that needs it, instead of running it globally inside of a `before` clause, so that it doesn't affect this test.